Core/include/Acts/Utilities/TrackHelpers.hpp

```diff
@@ -0,0 +1,339 @@
+                   freeVector.template segment<3>(eFreeDir0),
+                   BoundaryCheck(true), s_onSurfaceTolerance)
+        .closest();
+  };
+
+  if (strategy == TrackExtrapolationStrategy::first) {
+    ACTS_VERBOSE("looking for first track state");
+
+    auto first = findFirst();
+    if (!first.ok()) {
+      ACTS_ERROR("no first track state found");
+      return first.error();
+    }
+
+    SurfaceIntersection intersection = intersect(*first);
+    if (!intersection) {
+      ACTS_ERROR("no intersection found");
+      return Result<std::pair<TrackStateProxy, double>>::failure(
+          TrackExtrapolationError::ReferenceSurfaceUnreachable);
+    }
+
+    ACTS_VERBOSE("found intersection at " << intersection.pathLength());
+    return std::make_pair(*first, intersection.pathLength());
+  }
+
+  if (strategy == TrackExtrapolationStrategy::last) {
+    ACTS_VERBOSE("looking for last track state");
+
+    auto last = findLast();
+    if (!last.ok()) {
+      ACTS_ERROR("no last track state found");
+      return last.error();
+    }
+
+    SurfaceIntersection intersection = intersect(*last);
+    if (!intersection) {
+      ACTS_ERROR("no intersection found");
+      return Result<std::pair<TrackStateProxy, double>>::failure(
+          TrackExtrapolationError::ReferenceSurfaceUnreachable);
+    }
+
+    ACTS_VERBOSE("found intersection at " << intersection.pathLength());
+    return std::make_pair(*last, intersection.pathLength());
+  }
+
+  if (strategy == TrackExtrapolationStrategy::firstOrLast) {
+    ACTS_VERBOSE("looking for first or last track state");
+
+    auto first = findFirst();
+    if (!first.ok()) {
+      ACTS_ERROR("no first track state found");
+      return first.error();
+    }
+
+    auto last = findLast();
+    if (!last.ok()) {
+      ACTS_ERROR("no last track state found");
+      return last.error();
+    }
+
+    SurfaceIntersection intersectionFirst = intersect(*first);
+    SurfaceIntersection intersectionLast = intersect(*last);
+
+    double absDistanceFirst = std::abs(intersectionFirst.pathLength());
+    double absDistanceLast = std::abs(intersectionLast.pathLength());
+
+    if (intersectionFirst && absDistanceFirst <= absDistanceLast) {
+      ACTS_VERBOSE("using first track state with intersection at "
+                   << intersectionFirst.pathLength());
+      return std::make_pair(*first, intersectionFirst.pathLength());
+    }
+
+    if (intersectionLast && absDistanceLast <= absDistanceFirst) {
+      ACTS_VERBOSE("using last track state with intersection at "
+                   << intersectionLast.pathLength());
+      return std::make_pair(*last, intersectionLast.pathLength());
+    }
+
+    ACTS_ERROR("no intersection found");
+    return Result<std::pair<TrackStateProxy, double>>::failure(
+        TrackExtrapolationError::ReferenceSurfaceUnreachable);
+  }
+
+  throw std::runtime_error("invalid extrapolation strategy");
```

Comment:
Philosophical question if this should also use the return value.

Also (NIT): consider making this a switch over the enum instead of an if-chain.

Reply:
My favorite solution would be a compile time check if all cases are handled, but I think this is not possible in C++?

I kind of avoid switch statements for such big blocks of code but this is only a personal preference. Happy to change this if you think this aligns more with the rest of our repo

Reply:
Maybe this is what you meant, and I'm telling you what you well know, but in case it's useful: if you use a `switch` with no `default` and compile with `-Wall` (we do, don't we?), the compiler should tell you if you missed a `case`.

Reply:
but this is rather a compiler feature than a language feature right? not sure if we usually rely on this in acts

Reply:
GCC, Clang, and MSVC all have this warning. I think that is a fairly reliable compile-time check, even if it isn't (I believe) formally against the standard.